### PR TITLE
Add Sign-in component

### PR DESCRIPTION
**Pull request by 928PJY**

### CODE REVIEW

**Comment by fenxu** [comment]
>             "GitHubOauthScope": "repo,user:email read:user",

no need "," here? #Resolved

---
Refers to: configs/vscode-docs-build.json:5 in b93fac8. [](commit_id = b93fac80b435dd7c137ff0e9ffee5993b0d3a193, deletion_comment = False)

File: src/observers/BaseStatusBarObserver.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { StatusBarItem } from 'vscode';
+import { BaseEvent } from '../common/loggingEvents';
+
+export abstract class BaseStatusBarItemObserver {
```
Comment (fenxu) [inline src/observers/BaseStatusBarObserver.ts]: >BaseStatusBarItemObserver [](start = 22, length = 25)

Any other class besides SignStatusBarObeserver.ts will use this? #Closed

Comment (928PJY) [reply src/observers/BaseStatusBarObserver.ts]: there will be a buildStatusBarObserver in next PR

---
In reply to: [363139713](https://github.com/docascode/vscode-docs-build/pull/10#discussion_r363139713) [](ancestors = 363139713)

File: test/utils/faker.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { SinonSandbox } from 'sinon';
+import { EnvironmentController } from '../../src/common/EnvironmentController';
+
+export function MocPRODEnv(sinon: SinonSandbox, environmentController: EnvironmentController) {
```
Comment (live1206) [inline test/utils/faker.ts]: >MocPRODEnv [](start = 16, length = 10)

do you mean `Mock`? #Resolved

File: src/common/EventStream.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import { Subject, Subscription } from 'rxjs';
-import { BaseEvent } from "./loggingEvents";
+import { BaseEvent } from './loggingEvents';
 
 export class EventStream {
```
Comment (live1206) [inline src/common/EventStream.ts:4]: >EventStream [](start = 13, length = 11)

some .ts file has Camel case and some has Pascal case, can they align? #Resolved

**Review by partychen — APPROVED** [review]

File: src/common/eventType.ts [hunk 4]
```diff
@@ -1,4 +1,15 @@
 export enum EventType {
+    // Sign
+    CredentialInitializing,
```
Comment (live1206) [inline src/common/eventType.ts:4]: group with `Creadential` ones #Resolved

File: src/common/eventType.ts [hunk 5]
```diff
@@ -1,4 +1,15 @@
 export enum EventType {
+    // Sign
+    CredentialInitializing,
+    SignInProgress,
```
Comment (live1206) [inline src/common/eventType.ts]: >SignInProgress [](start = 4, length = 14)

`UserSignInProgress`? group with `User` ones #Resolved

File: src/credential/credentialController.ts [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    readonly aadInfo: string;
+    readonly userInfo: UserInfo;
+}
+
+export class CredentialController implements vscode.Disposable {
+    private signInStatus: DocsSignInStatus;
+    private aadInfo: string | undefined;
+    private userInfo: UserInfo | undefined;
+
+    public onDidChangeCredential: vscode.Event<Credential>;
```
Comment (live1206) [inline src/credential/credentialController.ts]: >onDidChangeCredential [](start = 11, length = 21)

what does `onDidChange` mean? #Resolved

Comment (928PJY) [reply src/credential/credentialController.ts]: removed, used before event driven framework.

---
In reply to: [363585798](https://github.com/docascode/vscode-docs-build/pull/10#discussion_r363585798) [](ancestors = 363585798)

File: src/credential/credentialController.ts [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+    constructor(private keyChain: KeyChain, private eventStream: EventStream, private environmentController: EnvironmentController) { }
+
+    public dispose(): void {
```
Comment (live1206) [inline src/credential/credentialController.ts]: >dispose [](start = 11, length = 7)

why empty `dispose`? #Resolved

Comment (928PJY) [reply src/credential/credentialController.ts]: can be removed, nothing need to be disposed

---
In reply to: [363593823](https://github.com/docascode/vscode-docs-build/pull/10#discussion_r363593823) [](ancestors = 363593823)

**Review by live1206 — APPROVED** [review]
:shipit:

**Review by fenxu — APPROVED** [review]
:shipit:

**Pull request merged**